File: core/vaultstore/src/test/kotlin/xyz/libravault/core/vaultstore/testing/FakeHardwareKeyWrap.kt
package xyz.libravault.core.vaultstore.testing

import xyz.libravault.core.vaultcrypto.VaultAuthenticationException
import xyz.libravault.core.vaultstore.HardwareKeyWrap
import xyz.libravault.core.vaultstore.HardwareKeyWrapFactory
import xyz.libravault.core.vaultstore.KeystoreHardwareUnavailableException
import xyz.libravault.core.vaultstore.KeystoreKeyLostException
import xyz.libravault.core.vaultstore.WrappedBlob
import java.security.SecureRandom
import javax.crypto.AEADBadTagException
import javax.crypto.BadPaddingException
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * JVM-only fake for [HardwareKeyWrapFactory] — this is what lets [VaultStore][
 * xyz.libravault.core.vaultstore.VaultStore]'s create/unlock/lock orchestration
 * be unit-tested without a device or Robolectric, per the repo lesson this
 * module was built to not repeat: `ProStateManager` (core:licensing) hardcodes
 * real Keystore access and is untestable in plain JVM tests as a result.
 *
 * Behaves like real hardware-backed Keystore keys for the properties tests
 * care about (deterministic per-alias key, AEAD wrap/unwrap, clean failure on
 * a wrong/missing key) without touching `android.security.keystore` at all.
 */
class FakeHardwareKeyWrapFactory : HardwareKeyWrapFactory {

    private val keysByAlias = mutableMapOf<String, ByteArray>()

    /** Set true to simulate a device with no hardware-backed Keystore at all —
     * exercises the [KeystoreHardwareUnavailableException] path. */
    var simulateHardwareUnavailable: Boolean = false

    override fun createNew(keyAlias: String): HardwareKeyWrap {
        if (simulateHardwareUnavailable) {
            throw KeystoreHardwareUnavailableException(IllegalStateException("fake: hardware unavailable"))
        }
        val key = ByteArray(32).also { SecureRandom().nextBytes(it) }
        keysByAlias[keyAlias] = key
        return FakeHardwareKeyWrap(key)
    }

    override fun forExisting(keyAlias: String): HardwareKeyWrap {
        val key = keysByAlias[keyAlias] ?: throw KeystoreKeyLostException(keyAlias)
        return FakeHardwareKeyWrap(key)
    }

    /** Simulates losing the Keystore key while the vault's files survive —
     * implementation plan §A.4 failure case (c), the scenario the recovery
     * key (§A.5) exists to rescue. */
    fun forgetKey(keyAlias: String) {
        keysByAlias.remove(keyAlias)
    }
}

private class FakeHardwareKeyWrap(private val key: ByteArray) : HardwareKeyWrap {

    override val isHardwareBacked: Boolean = true // fakes always report success; unavailability is a factory-level concern

    override fun wrap(plaintext: ByteArray): WrappedBlob {
        val nonce = ByteArray(12).also { SecureRandom().nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        return WrappedBlob(nonce, cipher.doFinal(plaintext))
    }

    override fun unwrap(wrapped: WrappedBlob): ByteArray {
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, wrapped.nonce))
        return try {
            cipher.doFinal(wrapped.ciphertext)
        } catch (e: AEADBadTagException) {
            throw VaultAuthenticationException(e)
        } catch (e: BadPaddingException) {
            throw VaultAuthenticationException(e)
        }
    }
}

File: ios/LibraVaultApp/LibraVault/LibraVaultTests/FakeHardwareKeyWrap.swift
import Foundation
@testable import LibraVault

/// Test-only fake for `HardwareKeyWrapFactory` — lets `VaultStore`'s create/
/// unlock/lock orchestration be unit-tested without a real Secure Enclave
/// (the Simulator has none — see `SecureEnclaveHardwareKeyWrapTests`'s own
/// doc comment). Mirrors Android's `FakeHardwareKeyWrapFactory`
/// (`core/vaultstore/testing`).
///
/// Behaves like a real hardware-backed key for the properties tests care
/// about (deterministic per-alias key, AEAD wrap/unwrap, clean failure on a
/// wrong/missing key) by reusing `AesGcmCipher`/`SecureRandom` directly
/// rather than going anywhere near `Security`/Secure Enclave APIs. Not the
/// same construction as the real `SecureEnclaveHardwareKeyWrap` (that one
/// uses ECIES; this uses plain AES-GCM with the nonce prefixed onto
/// `WrappedBlob.ciphertext`, since `WrappedBlob` has no separate nonce
/// field) — that's fine, this only needs to satisfy the `HardwareKeyWrap`
/// contract, not reproduce ECIES's own internals.
final class FakeHardwareKeyWrapFactory: HardwareKeyWrapFactory {

    private var keysByAlias: [String: Data] = [:]

    /// Set true to simulate a device with no Secure Enclave at all —
    /// exercises the `.secureEnclaveUnavailable` path.
    var simulateHardwareUnavailable = false

    func createNew(keyAlias: String) throws -> HardwareKeyWrap {
        if simulateHardwareUnavailable {
            throw HardwareKeyWrapError.secureEnclaveUnavailable(message: "fake: hardware unavailable")
        }
        let key = try SecureRandom.bytes(count: 32)
        keysByAlias[keyAlias] = key
        return FakeHardwareKeyWrap(key: key)
    }

    func forExisting(keyAlias: String) throws -> HardwareKeyWrap {
        guard let key = keysByAlias[keyAlias] else {
            throw HardwareKeyWrapError.keyLost(keyAlias: keyAlias)
        }
        return FakeHardwareKeyWrap(key: key)
    }

    /// Simulates losing the Secure Enclave key while the vault's files
    /// survive — the scenario the recovery key exists to rescue.
    func forgetKey(_ keyAlias: String) {
        keysByAlias.removeValue(forKey: keyAlias)
    }
}

private struct FakeHardwareKeyWrap: HardwareKeyWrap {
    let key: Data
    // Fakes always report success; unavailability is a factory-level concern.
    var isHardwareBacked: Bool { true }

    func wrap(_ plaintext: Data) throws -> WrappedBlob {
        let nonce = try SecureRandom.bytes(count: VaultFormat.nonceSizeBytes)
        let ciphertext = try AesGcmCipher.encrypt(key: key, nonce: nonce, aad: Data(), plaintext: plaintext)
        return WrappedBlob(ciphertext: nonce + ciphertext)
    }

    func unwrap(_ wrapped: WrappedBlob) throws -> Data {
        guard wrapped.ciphertext.count >= VaultFormat.nonceSizeBytes else {
            throw VaultCryptoError.authenticationFailed
        }
        let nonce = wrapped.ciphertext.prefix(VaultFormat.nonceSizeBytes)
        let ciphertext = wrapped.ciphertext.dropFirst(VaultFormat.nonceSizeBytes)
        return try AesGcmCipher.decrypt(key: key, nonce: nonce, aad: Data(), ciphertextWithTag: ciphertext)
    }
}
